src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/RenameTypeService.cs

```diff
@@ -0,0 +1,189 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Collections.Generic;
+using System.ComponentModel.Composition;
+using System.IO;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+
+using Microsoft.CodeAnalysis;
+using Microsoft.VisualStudio.LanguageServices;
+using Microsoft.VisualStudio.ProjectSystem.LanguageServices;
+
+namespace Microsoft.VisualStudio.ProjectSystem.VS
+{
+    [Export(typeof(IRenameTypeService))]
+    internal class RenameTypeService : IRenameTypeService
+    {
+        private readonly VisualStudioWorkspace _workspace;
+
+        [ImportingConstructor]
+        public RenameTypeService(
+            UnconfiguredProject project,
+            VisualStudioWorkspace visualStudioWorkspace)
+        {
+            _workspace = visualStudioWorkspace;
+            SyntaxFactsServicesImpl = new OrderPrecedenceImportCollection<ISyntaxFactsService>(projectCapabilityCheckProvider: project);
+        }
+
+        [ImportMany]
+        protected OrderPrecedenceImportCollection<ISyntaxFactsService> SyntaxFactsServicesImpl { get; }
+
+        private ISyntaxFactsService SyntaxFactsService => SyntaxFactsServicesImpl.First().Value;
+
+        public async Task<bool> DoesFileNameMatchTypeAsync(string filePath)
+        {
+            string name = Path.GetFileNameWithoutExtension(filePath);
+            ISymbol symbol = await TryGetSymbolToRenameAsync(name, filePath);
+            return symbol != null;
+        }
+
+        private string _expectedFilePath = null;
+        private static readonly object s_filePathLock = new object();
+
+        public void QueueFileWithTypeToBeRenamed(string filePath)
+        {
+            lock (s_filePathLock)
+            {
+                _expectedFilePath = filePath;
+            }
+        }
+
+        public async Task<bool> AnyTypeToRenameAsync(string oldFilePath, string newFilePath)
+        {
+            string expectedFilePath = _expectedFilePath;
+            if (expectedFilePath != oldFilePath)
+                return false;
+
+            string oldName = Path.GetFileNameWithoutExtension(oldFilePath);
+            string newName = Path.GetFileNameWithoutExtension(newFilePath);
+
+            if (!CanHandleRename(oldName, newName))
+                return false;
+
+            ISymbol symbol = await TryGetSymbolToRenameAsync(oldName, newFilePath);
```

Comment:
Why do we mix ISymbol and CodeModel?

Reply:
It feels like they can only get out of sync with regards to what they look for in terms of symbols, etc.